jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUTreeIsomorphismInspector.java

```diff
@@ -0,0 +1,349 @@
+import org.jgrapht.alg.util.Pair;
+import org.jgrapht.traverse.BreadthFirstIterator;
+import org.jgrapht.util.RadixSort;
+
+import java.util.*;
+
+/**
+ * This is an implementation of the AHU algorithm for detecting an (unweighted) isomorphism between two trees.
+ * Please see <a href="http://mathworld.wolfram.com/GraphIsomorphism.html">mathworld.wolfram.com</a> for a complete
+ * definition of the isomorphism problem for general graphs.
+ *
+ * <p>
+ *     The original algorithm was first presented in "Alfred V. Aho and John E. Hopcroft. 1974.
+ *     The Design and Analysis of Computer Algorithms (1st ed.). Addison-Wesley Longman Publishing Co., Inc., Boston, MA, USA."
+ * </p>
+ *
+ * <p>
+ *     This implementation runs in linear time (in the number of vertices of the input trees)
+ *     while using a linear amount of memory.
+ * </p>
+ *
+ * <p>
+ *      For an implementation that supports rooted forests see {@link AHUForestIsomorphismInspector}.
+ * </p>
+ *
+ * <p>
+ *     Note: This inspector only returns a single mapping (chosen arbitrarily) rather than all possible mappings.
+ * </p>
+ *
+ * @param <V> the type of the vertices
+ * @param <E> the type of the edges
+ *
+ * @author Alexandru Valeanu
+ */
+public class AHUTreeIsomorphismInspector<V, E> implements IsomorphismInspector<V, E> {
+    private final Graph<V, E> tree1;
+    private final Graph<V, E> tree2;
+
+    private V root1;
+    private V root2;
+
+    private Map<V, V> forwardMapping;
+    private Map<V, V> backwardMapping;
+
+    /**
+     * Construct a new AHU unrooted tree isomorphism inspector.
+     *
+     * Note: The constructor does NOT check if the input trees are valid.
+     *
+     * @param tree1 the first tree
+     * @param tree2 the second tree
+     * @throws NullPointerException if {@code tree1} is {@code null}
+     * @throws NullPointerException if {@code tree2} is {@code null}
+     */
+    public AHUTreeIsomorphismInspector(Graph<V, E> tree1, Graph<V, E> tree2){
+        this.tree1 = Objects.requireNonNull(tree1, "tree1 cannot be null");
+        this.tree2 = Objects.requireNonNull(tree2, "tree2 cannot be null");
+    }
+
+    /**
+     * Construct a new AHU rooted tree isomorphism inspector.
+     *
+     * Note: The constructor does NOT check if the input trees are valid.
+     *
+     * @param tree1 the first rooted tree
+     * @param root1 the root of the first tree
+     * @param tree2 the second rooted tree
+     * @param root2 the root of the second tree
+     * @throws NullPointerException if {@code tree1} is {@code null}
+     * @throws NullPointerException if {@code root1} is {@code null}
+     * @throws NullPointerException if {@code tree2} is {@code null}
+     * @throws NullPointerException if {@code root2} is {@code null}
+     * @throws IllegalArgumentException if either {@code root1} or {@code root2} contain an invalid vertex
+     */
+    public AHUTreeIsomorphismInspector(Graph<V, E> tree1, V root1, Graph<V, E> tree2, V root2){
```

Comment:
Change the functionality to allow null root values, so that the other constructor can call this more generic constructor.